File: src/Functional/Clojure/chain_of_responsibility.clj
(require '[clojure.string :as str])

(def handlers
  [{:name "faq" :accepts #(<= % 50)}
   {:name "billing" :accepts #(<= % 500)}
   {:name "escalation" :accepts (constantly true)}])

(defn route-request [amount chain]
  (loop [remaining chain
         visited []]
    (if-let [handler (first remaining)]
      (let [visited-now (conj visited (:name handler))]
        (if ((:accepts handler) amount)
          [visited-now (:name handler)]
          (recur (rest remaining) visited-now)))
      (throw (ex-info "No handler accepted the request" {:amount amount})))))

(let [amount 250
      [visited handled] (route-request amount handlers)]
  (println
    (format "visited=%s;handled=%s;result=refund(%d)"
            (str/join ">" visited)
            handled
            amount)))

File: src/Functional/OCaml/chain_of_responsibility.ml
type handler = {
  name : string;
  accepts : int -> bool;
}

let rec route amount visited = function
  | [] -> failwith "No handler accepted the request"
  | handler :: rest ->
      let visited_now = handler.name :: visited in
      if handler.accepts amount then
        (List.rev visited_now, handler.name)
      else
        route amount visited_now rest

let handlers = [
  { name = "faq"; accepts = (fun amount -> amount <= 50) };
  { name = "billing"; accepts = (fun amount -> amount <= 500) };
  { name = "escalation"; accepts = (fun _ -> true) };
]

let () =
  let amount = 250 in
  let visited, handled = route amount [] handlers in
  Printf.printf "visited=%s;handled=%s;result=refund(%d)\n"
    (String.concat ">" visited) handled amount
